src/Microsoft.Azure.ServiceBus/ServiceBusDiagnosticsSource.cs

```diff
@@ -0,0 +1,771 @@
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    Session = session,
+                    Message = message,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region Schedule
+
+        internal Activity ScheduleStart(Message message, DateTimeOffset scheduleEnqueueTimeUtc)
+        {
+            Activity activity = Start("Schedule", () => new
+                {
+                    Message = message,
+                    ScheduleEnqueueTimeUtc = scheduleEnqueueTimeUtc,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint
+                },
+                a => SetTags(a, message));
+
+            Inject(message);
+
+            return activity;
+        }
+
+        internal void ScheduleStop(Activity activity, Message message, DateTimeOffset scheduleEnqueueTimeUtc, TaskStatus? status, long sequenceNumber)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    Message = message,
+                    ScheduleEnqueueTimeUtc = scheduleEnqueueTimeUtc,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    SequenceNumber = sequenceNumber,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region Cancel
+
+        internal Activity CancelStart(long sequenceNumber)
+        {
+            return Start("Cancel", () => new
+            {
+                SequenceNumber = sequenceNumber,
+                Entity = this.entityPath,
+                Endpoint = this.endpoint
+            },
+            null);
+        }
+
+        internal void CancelStop(Activity activity, long sequenceNumber, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    SequenceNumber = sequenceNumber,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region Receive
+
+        internal Activity ReceiveStart(int messageCount)
+        {
+            return Start("Receive", () => new
+            {
+                MessageCount = messageCount,
+                Entity = this.entityPath,
+                Endpoint = this.endpoint
+            },
+            null);
+        }
+
+        internal void ReceiveStop(Activity activity, int messageCount, TaskStatus? status, IList<Message> messageList)
```

Comment:
`messageCount` is problematic. It represents the maximum number of messages ASB client _might_ fetch, but not necessarily the actual number. The actul number of retrieved messages should be retrieved from `messageList.Length` to be accurate.

Reply:
That's a good point. However image you are logging this event:
`Requested {messageCount}, received {messageList.Count} in a {Activity.Current.Duration}`.
I think it's useful to know how much did you request and how much did you receive in response.

As you notice, every 'activity' has a Start and Stop events and `messageCount` really makes sense in Start event payload ('Requesting {messageCount} messages'). However Start event is not really interesting, at least not as interesting as Stop event (when you know the result and duration). So we expect the majority of listeners to care about Stop events only (and reduce verbosity by disabling Start events), so we try to provide full context about every operation in Stop.

Reply:
Agreed. Therefore [`MessageCount`](https://github.com/Azure/azure-service-bus-dotnet/pull/370/files/8000966fc57dde2106488864075c512213abb084#diff-b0e75e823e0fccc7d74474f07ab71968R229) should not be assigned the value of `messageCount` but `messageList.Length` instead.

Reply:
I'm sorry, I did not understand it.
If I never received Start event (and this is a default case) and on Stop event I want to log 
`Requested {messageCount}, received {messageList.Count} in a {Activity.Current.Duration}.`
How would I do that if `messageCount` is `messageList.Count`? 

Would you be ok with renaming `MessageCount` to e.g. `RequestedMessageCount`?